stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/Channel.java

```diff
@@ -39,50 +43,154 @@
+     */
+    private Reader reader;
+
+    /**
+     * Specifies how the content should be written to a channel.
+     */
+    private Writer writer;
+
+    /**
+     * Specifies whether the channel has reached EoF.
+     */
+    private boolean hasReachedToEnd = false;
 
     /**
      * Holds the content belonging to a particular channel.
      */
     private Buffer contentBuffer;
 
+    private static final Logger log = LoggerFactory.getLogger(Channel.class);
+
     /**
      * <p>
-     * Specifies the maximum number of bytes which will be read in each chunk.
+     * Will read/writ bytes from the provided channel
      * </p>
      * <p>
-     * This value will be used when reading all the bytes from the channel.
+     * This operation will asynchronously read data from the channel.
      * </p>
      *
-     * @see Channel#readAll()
+     * @param channel which will be used to read/write content.
+     * @param reader  will be used for reading content.
+     * @param writer  will be used for writing content.
+     * @throws BallerinaIOException initialization error.
      */
-    private static final int MAX_BUFFER_CHUNK_SIZE = 1024;
+    public Channel(ByteChannel channel, Reader reader, Writer writer) throws BallerinaIOException {
+        if (null != channel) {
+            this.channel = channel;
+            this.reader = reader;
+            this.writer = writer;
+            if (log.isDebugEnabled()) {
+                log.debug("Initializing ByteChannel with ref id " + channel.hashCode());
+            }
+        } else {
+            String message = "Provided channel cannot be initialized";
+            throw new BallerinaIOException(message);
+        }
+    }
 
-    private static final Logger log = LoggerFactory.getLogger(Channel.class);
+    /**
+     * <p>
+     * Creates a channel which will contain a fixed sized buffer.
+     * </p>
+     * <p>
+     * This operation will in turn request for excess bytes than required. And buffer the content. Also this will
+     * mostly be used with blocking readers and writers.
+     * </p>
+     *
+     * @param channel which will be used to read/write content.
+     * @param reader  will be used for reading content.
+     * @param writer  will be used for writing content.
+     * @param size    the size of the fixed buffer.
+     * @throws BallerinaIOException initialization error.
+     */
+    public Channel(ByteChannel channel, Reader reader, Writer writer, int size) throws BallerinaIOException {
+        if (null != channel) {
+            this.channel = channel;
+            this.reader = reader;
+            this.writer = writer;
+            contentBuffer = new Buffer(size);
+            if (log.isDebugEnabled()) {
+                log.debug("Initializing ByteChannel with ref id " + channel.hashCode());
+            }
+        } else {
+            String message = "Provided channel cannot be initialized";
+            throw new BallerinaIOException(message);
```

Comment:
Shall we convert this to a single line?

Reply:
This way is more readable and extensible IMO. Let's say we want to log the message. If so instead of duplicating we could use the message itself